elastic-apm-agent/src/main/java/co/elastic/apm/agent/premain/ShadedClassLoader.java

```diff
@@ -80,23 +101,39 @@ public ShadedClassLoader(File jar, ClassLoader parent, String customPrefix) thro
 
     private Class<?> defineClass(String name, byte[] classBytes) {
         String packageName = getPackageName(name);
-        if (packageName != null && getPackage(packageName) == null) {
+        if (packageName != null && isPackageNotDefined(packageName)) {
             try {
                 if (manifest != null) {
-                    definePackage(name, manifest, jarUrl);
+                    definePackage(packageName, manifest, jarUrl);
                 } else {
                     definePackage(packageName, null, null, null, null, null, null, null);
                 }
-            } catch (IllegalArgumentException e) {
+            } catch (IllegalArgumentException e){
                 // The package may have been defined by a parent class loader in the meantime
-                if (getPackage(packageName) == null) {
+                if (isPackageNotDefined(packageName)) {
                     throw e;
                 }
             }
         }
         return defineClass(name, classBytes, 0, classBytes.length, ShadedClassLoader.class.getProtectionDomain());
     }
 
+    @SuppressWarnings("deprecation")
+    private boolean isPackageNotDefined(String packageName) {
+        Package pkg;
+        if (getDefinedPackage != null) {
+            try {
+                pkg = (Package) getDefinedPackage.invokeExact(this, packageName);
```

Comment:
Do we even need to call `getDefinedPackage` or is `getPackage` good enough? Maybe its preferable to have the same semantics regardless of the Java version?

On Slack you said

> f migrating the call to the deprecated `getPackage(...)` and using `getDefinedPackage(...)` the difference is that now it does not tries to delegate to parent nor Bootstrap CL when checking if the package is defined.
Given the classes that are loaded by the shaded CL can only be loaded by this CL, I don't think there is any need to ask the parent CL in this case.

So it seems there's not really a benefit of calling `getDefinedPackage`?

Reply:
Ah, I misunderstood. The `getPackage` delegates to the parent which may be an issue in certain cases if the package has been defined in a parent CL but not the current CL. Probably not an issue either way in our case, as the parent (platform CL or bootstrap CL) will never have defined an agent package.

I'm fine either way, I guess. Calling `getDefinedPackage` is technically more correct and avoids calling a deprecated API but adds a bit of boiler plate.

Reply:
That's a very good question !

In theory there is no case where the `ShadedClassLoader` will load classes that are defined in packages that have been defined by a parent classloader, so using `getDefinedPackage` should be enough and we can skip the extra lookups that are done by `getPackage`. Performance wise, it should be a trivial difference as the packages are stored in a concurrent map.

In practice, we still have cases where the call the `defineClass` throws an `IllegalArgumentException`, which means a parent CL is expected to load those classes. So if the classes can be defined elsewhere, then it would also be consistent to consider that the packages can also be defined elsewhere, hence delegation to parent CL does make sense.
On top of that, we only check for the package presence with a `null` check and don't use the returned `Package` instance, so there is no difference on the origin of the `Package` instance.

As a conclusion I now think it's not worth the extra complexity to use a method handle here, and we can always revisit this later. I'll add a comment to clarify the usage of `getPackage` vs `getDefinedPackage`.